[WK2] Provide IPC encoding, decoding helper methods for objects

zdobersek opened the pull request:
#### 4e2a24122f1668f09693e0df1510bd07da46fe46
<pre>
[WK2] Provide IPC encoding, decoding helper methods for objects
<a href="https://bugs.webkit.org/show_bug.cgi?id=250785">https://bugs.webkit.org/show_bug.cgi?id=250785</a>

Reviewed by Kimmo Kinnunen.

Provide helper methods for encoding trivially-copyable objects on the IPC
encoder and decoder classes. These would be used instead of encoding and
decoding these objects as fixed-length data. That would of course still be done
internally, but through a nicer interface.

encodeObject() on the IPC::Encoder and IPC::StreamConnectionEncoder classes
takes a given object of a trivially-copyable type and encodes it as fixed-length
data through a single-element Span. decodeObject() on the IPC::Decoder class
reverses that, decoding a single-element Span and returning an object of a
trivially-copyable type.

Remaining uses of encodeFixedLengthData() and decodeFixedLengthData() methods
are replaced with the new alternatives.

* Source/WTF/wtf/ArgumentCoder.h:
(IPC::ArgumentCoder&lt;bool&gt;::decode):
* Source/WebCore/platform/audio/cocoa/CAAudioStreamDescription.h:
(WebCore::CAAudioStreamDescription::encode const):
(WebCore::CAAudioStreamDescription::decode):
* Source/WebCore/platform/mediastream/RealtimeMediaSourceCapabilities.h:
(WebCore::CapabilityValueOrRange::encode const):
(WebCore::CapabilityValueOrRange::decode):
* Source/WebKit/Platform/IPC/ArgumentCoders.h:
(IPC::SimpleArgumentCoder::encode):
(IPC::SimpleArgumentCoder::decode):
* Source/WebKit/Platform/IPC/Decoder.h:
(IPC::Decoder::decodeObject):
* Source/WebKit/Platform/IPC/Encoder.h:
(IPC::Encoder::encodeObject):
* Source/WebKit/Platform/IPC/StreamConnectionEncoder.h:
* Source/WebKit/Shared/WebCoreArgumentCoders.cpp:
(IPC::ArgumentCoder&lt;RectEdges&lt;bool&gt;&gt;::decode):
(IPC::ArgumentCoder&lt;ViewportArguments&gt;::decode):
* Source/WebKit/Shared/WebCoreArgumentCoders.h:
* Source/WebKit/Shared/win/WebCoreArgumentCodersWin.cpp:
(IPC::ArgumentCoder&lt;LOGFONT&gt;::encode):
(IPC::ArgumentCoder&lt;LOGFONT&gt;::decode):

Canonical link: <a href="https://commits.webkit.org/259638@main">https://commits.webkit.org/259638@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/1f405164a7daa15ace4e4e4ffebaf901470d31f4

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/105472 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/14580 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/38356 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/114742 "Built successfully") | [  ~~🛠 🧪 win~~](https://ews-build.webkit.org/#/builders/10/builds/174885 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/109373 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/15868 "Built successfully") | [  ~~🛠 mac-AS-debug~~](https://ews-build.webkit.org/#/builders/85/builds/5783 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/97777 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/12/builds/114584 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/111228 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/12162 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/95148 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/35/builds/39651 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/94027 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/26783 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/81336 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [  ~~🛠 🧪 jsc~~](https://ews-build.webkit.org/#/builders/1/builds/95250 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/7856 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/28139 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/38/builds/93368 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/83/builds/5631 "Built successfully and passed tests") | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/82/builds/8215 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/85/builds/5783 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/46/builds/30375 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/13987 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/47682 "Passed tests") | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/37/builds/102075 "Built successfully") | 
| [❌ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/6649 "Failed to push commit to Webkit repository") | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/79/builds/9883 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/45/builds/25457 "Passed tests") | 
| | | | | 
<!--EWS-Status-Bubble-End-->

achristensen07 commented on Source/WebKit/Platform/IPC/Decoder.h:
```diff
@@ -95,6 +95,9 @@ class Decoder {
     WARN_UNUSED_RETURN Span<const T> decodeSpan(size_t);
 
+    template<typename T>
+    WARN_UNUSED_RETURN bool decodeObject(T&);
```
std::optional<T> decodeObject()

achristensen07 reviewed (commented):
Why did you write this?  I feel like this isn't heading in a good direction.

kkinnunen-apple commented on Source/WebCore/platform/mediastream/RealtimeMediaSourceCapabilities.h:
```diff
@@ -114,17 +114,30 @@ class CapabilityValueOrRange {
+    auto type = decoder.template decode<Type>();
+    if (!type)
+        return false;
+
```
The pattern for the decode should be:
```
auto minOrValue = decoder.template decodeObject<ValueUnion>();
auto max = decoder.template decodeObject<ValueUnion>();
auto type = decoder.template decode<Type>();
if (UNLIKELY(!decoder.isValid())
    return false;
```

E.g. the common case be "branch free" from the caller statement list.

kkinnunen-apple replied:
With respect to the comment above about matching ArgumentCoders to implementations, it'd be good to regularise this eventually to
```
auto minOrValue = decoder.template decode<ValueUnion>();
auto max = decoder.template decode<ValueUnion>();
auto type = decoder.template decode<Type>();
if (UNLIKELY(!decoder.isValid())
    return false;
```

With respect to comment below about legacy decoder, this markup would be
```
(void) decoder.decode(&m_minOrValue);
(void) decoder.decode(&m_max);
(void  decoder.decode(&m_type);
if (UNLIKELY(!decoder.isValid())
    return false;
return true;
```
However I don't think this is that important at this point.

zdobersek replied:
I think applying batched, branch-free decoding will be simplest (or already is simplest) in the generated serializers code.

Decoding from an already-invalid decoder perhaps deserves some investigation. The buffer position would be null, and I believe any decoding would fail early in the buffer containment check, but I didn't review thoroughly.

kkinnunen-apple commented on Source/WebKit/Platform/IPC/Decoder.h:
```diff
@@ -244,4 +247,18 @@ inline Span<const T> Decoder::decodeSpan(size_t size)
+    memcpy(&object, data.data(), sizeof(T));
+    return object;
+}
+
```
Since the decodeObject is called currently in the legacy decoder pattern, one could envision 
`bool Decoder::decodeObject(T&)` be faster, so one could write that variant for the time being.  But we don't have the benchmark for that, so probably currently it does not matter. And as such, this comment is just a FYI, not needed for this patch.

zdobersek replied:
I compared generated code for `bool Decoder::decodeObject(T&)` and `std::optional<T> Decoder::decodeObject()` variants, and they're practically identical, with the std::optional value fitting into one return register and the charged value fitting into the other one (speaking for register-fitting values). There's going to be subtle differences in code handling the returned values, but I don't think that difference is substantial.

kkinnunen-apple commented on Source/WebKit/Platform/IPC/Decoder.h:
```diff
@@ -244,4 +247,18 @@ inline Span<const T> Decoder::decodeSpan(size_t size)
+    if (!data.data())
+        return std::nullopt;
+
+    T object { };
```
so here your `data`  is intended to be "business as usual array of `T` of size 1."

In order to copy that into a optional return value, you do:
`return data[0];`

zdobersek replied:
Should probably work.

kkinnunen-apple commented on Source/WebCore/platform/audio/cocoa/CAAudioStreamDescription.h:
```diff
@@ -80,16 +80,16 @@ class WEBCORE_EXPORT CAAudioStreamDescription final : public AudioStreamDescript
 void CAAudioStreamDescription::encode(Encoder& encoder) const
 {
-    encoder.encodeFixedLengthData(reinterpret_cast<const uint8_t*>(&m_streamDescription), sizeof(m_streamDescription), 1);
+    encoder.encodeObject(m_streamDescription);
```
I think in medium term we want to remove the invocations of `encodeObject`. 
So what I think Alex was saying is that we don't want to have future where the per-member encode / decode logic sometimes calls "encode" and sometimes "encodeObject". For the most part, we want to have regular form.

For purpose of this discussion, encoding / decoding could be divided into three categories:

1. is_trivially_constructible argument coders
2. is_trivially_constructible argument coders that validate
3. per-member argument coders that may or may not validate

For current code that ends up using encodeObject / decodeObject in the 1. category, near term we should instead add a trivially copiable argument coder and then use the regular form (encode <<, decode<>()).

I don't think we have code in category 2. If we did, encodeObject/decodeObject would be a syntactic sugar, less error prone version over encodeSpan/decodeSpan. So in this light, it seems ok to me to add these in this patch since you've written it already. However, if you modify the patch according to the feedback below, then maybe we could try to live without encodeObject/decodeObject with the rationale of not adding yet yet another encode / decode method.


So here we should mark up ideally in this way:

1. AudioStreamBasicDescription is serialised with a trivially copiable argument coder
2. CAAudioStreamDescription is serialised with a member wise argument coder

So we are lacking 1. currently. If we had that, we wouldn't need to call encodeObject.
So we would do:

This would go to somewhere sensible, maybe `WebCoreArgumentEncodersCocoa.h`. The generated serializers don't seem to be able to generate this yet.

```
template<> struct ArgumentCoder<AudioStreamBasicDescription> : SimpleArgumentCoder< AudioStreamBasicDescription> { };
```

For the CAAudioStreamDescription, we could then have a regular by-member encode / decode.
Then in future patch, we would move the CAAudioStreamDescription in some generated serializer .in file.

zdobersek replied:
The plan was to take SimpleArgumentCoder, probably bring it into WTF's ArgumentCoder.h as something like `TriviallyCopyableArgumentCoder`, and provide ArgumentCoder specializations for AudioStreamBasicDescription as well as the union type below and I guess also LOGFONT on Windows. That brings everything back into uniformity.

kkinnunen-apple commented on Source/WebKit/Shared/win/WebCoreArgumentCodersWin.cpp:
```diff
@@ -39,14 +39,11 @@ using namespace WebCore;
     static void encode(Encoder& encoder, const LOGFONT& logFont)
     {
-        encoder.encodeFixedLengthData(reinterpret_cast<const uint8_t*>(&logFont), sizeof logFont, 1);
+        encoder.encodeObject(logFont);
```
Note: this changes the alignment of the encoding, which I think is ok.

kkinnunen-apple commented on Source/WebKit/Shared/WebCoreArgumentCoders.cpp:
```diff
@@ -262,9 +262,9 @@ void ArgumentCoder<RectEdges<bool>>::encode(Encoder& encoder, const RectEdges<bo
 }
     
-bool ArgumentCoder<RectEdges<bool>>::decode(Decoder& decoder, RectEdges<bool>& boxEdges)
+std::optional<RectEdges<bool>> ArgumentCoder<RectEdges<bool>>::decode(Decoder& decoder)
```
Note (not related to your patch):
since we are changing this, we could just change this to the correct simpleargumentcoder form, e.g. inheriting from it.

Also Note (not related to your patch):
this encodes bools as data, which highlights the futility of !!bool thing in bool encoder that makes the bool encoder more complex than it needs.

kkinnunen-apple commented on Source/WTF/wtf/ArgumentCoder.h:
```diff
@@ -92,9 +92,9 @@ struct ArgumentCoder<bool> {
     template<typename Decoder>
     static std::optional<bool> decode(Decoder& decoder)
```
Note (not related to your patch per se):
this !!bool pattern seems futile, as other bools inside trivially constructible structs are being passed as data.
so individual bools are "sanitized" as 1, but bools inside data are not. Which makes this one seem pointless.

zdobersek replied:
I'll apply this recommendation later.

kkinnunen-apple reviewed (approved):
I think this is a progression over the current one.
Some ideas that you may or may not incorporate in this patch. They should eventually be done.

zdobersek committed:
[WK2] Provide IPC encoding, decoding helper methods for objects
https://bugs.webkit.org/show_bug.cgi?id=250785

Reviewed by Kimmo Kinnunen.

Provide helper methods for encoding trivially-copyable objects on the IPC
encoder and decoder classes. These would be used instead of encoding and
decoding these objects as fixed-length data. That would of course still be done
internally, but through a nicer interface.

encodeObject() on the IPC::Encoder and IPC::StreamConnectionEncoder classes
takes a given object of a trivially-copyable type and encodes it as fixed-length
data through a single-element Span. decodeObject() on the IPC::Decoder class
reverses that, decoding a single-element Span and returning an object of a
trivially-copyable type.

Remaining uses of encodeFixedLengthData() and decodeFixedLengthData() methods
are replaced with the new alternatives.

* Source/WTF/wtf/ArgumentCoder.h:
(IPC::ArgumentCoder<bool>::decode):
* Source/WebCore/platform/audio/cocoa/CAAudioStreamDescription.h:
(WebCore::CAAudioStreamDescription::encode const):
(WebCore::CAAudioStreamDescription::decode):
* Source/WebCore/platform/mediastream/RealtimeMediaSourceCapabilities.h:
(WebCore::CapabilityValueOrRange::encode const):
(WebCore::CapabilityValueOrRange::decode):
* Source/WebKit/Platform/IPC/ArgumentCoders.h:
(IPC::SimpleArgumentCoder::encode):
(IPC::SimpleArgumentCoder::decode):
* Source/WebKit/Platform/IPC/Decoder.h:
(IPC::Decoder::decodeObject):
* Source/WebKit/Platform/IPC/Encoder.h:
(IPC::Encoder::encodeObject):
* Source/WebKit/Platform/IPC/StreamConnectionEncoder.h:
* Source/WebKit/Shared/WebCoreArgumentCoders.cpp:
(IPC::ArgumentCoder<RectEdges<bool>>::decode):
(IPC::ArgumentCoder<ViewportArguments>::decode):
* Source/WebKit/Shared/WebCoreArgumentCoders.h:
* Source/WebKit/Shared/win/WebCoreArgumentCodersWin.cpp:
(IPC::ArgumentCoder<LOGFONT>::encode):
(IPC::ArgumentCoder<LOGFONT>::decode):

Canonical link: https://commits.webkit.org/259638@main
